app/src/main/java/io/mochahub/powermeter/workoutsession/workoutsessiondialog/NewWorkoutDialog.kt

```diff
@@ -0,0 +1,340 @@
+package io.mochahub.powermeter.workoutsession.workoutsessiondialog
+
+import android.app.DatePickerDialog
+import android.content.DialogInterface
+import android.content.res.Configuration
+import android.content.res.Resources
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.RectF
+import android.os.Bundle
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.DialogFragment
+import androidx.navigation.fragment.navArgs
+import androidx.recyclerview.widget.ItemTouchHelper
+import com.airbnb.epoxy.EpoxyTouchHelper
+import io.mochahub.powermeter.R
+import io.mochahub.powermeter.data.AppDatabase
+import io.mochahub.powermeter.data.ExerciseEntity
+import io.mochahub.powermeter.models.Exercise
+import io.mochahub.powermeter.models.Workout
+import io.mochahub.powermeter.models.WorkoutSession
+import io.mochahub.powermeter.models.WorkoutSet
+import io.mochahub.powermeter.models.addSet
+import io.mochahub.powermeter.models.removeSet
+import io.mochahub.powermeter.models.setReps
+import io.mochahub.powermeter.models.setWeight
+import io.mochahub.powermeter.models.updateExercise
+import kotlinx.android.synthetic.main.dialog_new_workout.addWorkoutBtn
+import kotlinx.android.synthetic.main.dialog_new_workout.newWorkoutDateText
+import kotlinx.android.synthetic.main.dialog_new_workout.newWorkoutNameText
+import kotlinx.android.synthetic.main.dialog_new_workout.newWorkoutToolbar
+import kotlinx.android.synthetic.main.fragment_exercise.recyclerView
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.time.Instant
+import java.util.Calendar
+import java.util.Date
+import kotlin.math.absoluteValue
+
+class NewWorkoutDialog : WorkoutController.AdapterCallbacks, DialogFragment() {
+
+    private val args: NewWorkoutDialogArgs by navArgs()
+    private var shouldSave = true
+    private lateinit var workoutController: WorkoutController
+    private lateinit var viewModel: NewWorkoutViewModel
+    private var exercises = listOf<ExerciseEntity>()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setStyle(
+            STYLE_NORMAL,
+            R.style.FullScreenDialog
+        )
+
+        viewModel =
+            NewWorkoutViewModel(
+                db = AppDatabase(requireContext())
+            )
+        CoroutineScope(Dispatchers.IO).launch {
```

Comment:
Every instance of `CoroutineScope(Dispatchers.IO).launch {}` in `NewWorkoutDialog` should probably be taking place in the viewModel with `viewModelScope.launch {}`

Reply:
I agree I want to move them out of the dialog. Removed most of these instances, however, some of them can't be removed until the room relations are done correctly, which I think I should leave for a new PR. Leaving TODO's where this will be affected.